### PR TITLE
Add a C re-entrant critical section API

### PR DESCRIPTION
**Pull request by bremoran**
Implements the API discussed in https://github.com/ARMmbed/mbed-util/issues/16

### CODE REVIEW

File: source/critical.c [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    if (!interruptEnableCounter) {
+        critical_primask = __get_PRIMASK();
+    }
+    __disable_irq();
+    /* not allowed to overflow the interruptEnableCounter. */
+    if (interruptEnableCounter == UINT32_MAX) {
+        /* Generate a fault */
+        ((void(*)(void))&invalid)();
+    }
+    interruptEnableCounter++;
```
Comment (bremoran) [inline source/critical.c]: This should probably use atomic_incr

File: source/critical.c [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+void critical_section_exit()
+{
+    if (interruptEnableCounter) interruptEnableCounter--;
```
Comment (bremoran) [inline source/critical.c]: This should probably use atomic_decr

File: source/critical.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#include "mbed-util/critical.h"
+
+static volatile uint32_t interruptEnableCounter = 0;
+static volatile uint32_t critical_primask = 0;
+const uint32_t invalid = 0xFFFFFFFEUL;
+void critical_section_enter()
+{
+    /* sample the primask */
+    if (!interruptEnableCounter) {
+        critical_primask = __get_PRIMASK();
```
Comment (bremoran) [inline source/critical.c]: This should probably use atomic_cas

File: source/critical.c [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        /* Generate a fault */
+        ((void(*)(void))&invalid)();
+    }
+    interruptEnableCounter++;
+}
+
+void critical_section_exit()
+{
+    if (interruptEnableCounter) interruptEnableCounter--;
+    if (!interruptEnableCounter && !critical_primask) __enable_irq();
```
Comment (bremoran) [inline source/critical.c]: This should use `__set_PRIMASK`

**Comment by bremoran** [comment]
We should check for a race between `__get_PRIMASK` and `__disable_irq`

File: source/critical.c (hunk 3 above)
Comment (bremoran) [inline source/critical.c]: Reading the primask introduces a race condition. We should ignore the primask.

File: source/critical.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ */
+#include <stdint.h>                  // uint32_t, UINT32_MAX
+#include <stddef.h>                  // NULL
+#include "cmsis-core/core_generic.h" //__disable_irq, __enable_irq
+
+// Module include
+#include "mbed-util/critical.h"
+
+static volatile uint32_t interruptEnableCounter = 0;
+static volatile uint32_t critical_primask = 0;
```
Comment (bremoran) [inline source/critical.c]: This variable is not required.

File: source/critical.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+const uint32_t invalid = 0xFFFFFFFEUL;
+void critical_section_enter()
+{
+    /* sample the primask */
+    if (!interruptEnableCounter) {
+        critical_primask = __get_PRIMASK();
+    }
```
Comment (bremoran) [inline source/critical.c]: If interruptEnableCounter is non-zero, we should check the primask. If interrupts are enabled, we should assert that something has enabled them.

File: source/critical.c (hunk 4 above)
Comment (bremoran) [inline source/critical.c]: Actually, let's not.

File: source/critical.c (hunk 2 above)
Comment (bremoran) [inline source/critical.c]: We should check the primask here and see if someone has enabled interrupts when they shouldn't be.

File: source/critical.c (hunk 3 above)
Comment (adbridge) [inline source/critical.c]: Isn't atomic_cas c++ and thus not usable in a c based critical section lock function?

**Comment by 0xc0170** [comment]
I believe this one can be closed, resolved in #88

**Pull request closed without merging**